mrjob/dataproc.py

```diff
@@ -0,0 +1,873 @@
+        args.extend(self._hadoop_args_for_step(step_num))
+
+        mapper, combiner, reducer = (self._hadoop_streaming_commands(step_num))
+
+        if mapper:
+            args += ['-mapper', mapper]
+
+        if combiner:
+            args += ['-combiner', combiner]
+
+        if reducer:
+            args += ['-reducer', reducer]
+
+        for current_input_uri in self._step_input_uris(step_num):
+            args += ['-input', current_input_uri]
+
+        args += ['-output', self._step_output_uri(step_num)]
+
+        # TODO - mtai @ davidmarin - Add back support to specify a different jar URI
+        output_hadoop_job = {
+            "args": args,
+            "fileUris": file_uris,
+            "archiveUris": archive_uris,
+            "properties": properties,
+            "mainJarFileUri": main_jar_uri
+        }
+        return output_hadoop_job
+
+    def _launch_dataproc_cluster(self):
+        """Create an empty cluster on Dataproc, and set self._gcp_cluster to
+        its ID."""
+        self._wait_for_gcs_eventual_consistency()
+
+        self._gcp_cluster_start = time.time()
+
+        # make sure we can see the files we copied to GCS
+        log.info('Creating Dataproc Hadoop cluster')
+
+        # NOTE - mtai @ davidmarin - We create the cluster if its missing, otherwise we join an existing one
+        # NOTE - mtai @ davidmarin - The cluster MUST be created before submitting jobs, otherwise job submission fails
+
+        # TODO - mtai @ davidmarin - change the way we auto-create cluster names
+        # "clusterName must be a match of regex '(?:[a-z](?:[-a-z0-9]{0,53}[a-z0-9])?).'"
+        self._gcp_cluster = self._gcp_cluster or "default-%s" % self._opts['owner']
+
+        try:
+            self._get_cluster()
+            log.info('Adding our job to existing cluster %s' % self._gcp_cluster)
+        # TODO - mtai @ davidmarin - Change Exception to HttpError and check 404 for cluster does not exist
+        except Exception as e:
+            self._create_cluster()
+            log.info('Created new cluster %s' % self._gcp_cluster)
+
+        # keep track of when we launched our job
+        self._dataproc_job_start = time.time()
+
+    def _wait_for_steps_to_complete(self):
+        """Wait for every step of the job to complete, one by one."""
+
+        total_steps = self._num_steps()
+        # define out steps
+        for step_num in xrange(total_steps):
+            # NOTE - mtai @ davidmarin - Dataproc accepts multiple jobs at once
+            # NOTE - mtai @ davidmarin - In this particular implementation, we must build each step on the fly and submit it and WAIT for step completion
+            # NOTE - mtai @ davidmarin - we THEN iterate through the loop and continue on
+
+            # Build each step
+            hadoop_job = self._build_dataproc_hadoop_job(step_num)
+
+            # Clean-up step name
+            step_name = '%s - Step %d of %d' % (self._job_key, step_num + 1, self._num_steps())
```

Comment:
It seems a little ridiculous to create spaces and uppercase letters only to convert them to hyphens and lowercase letters. :)

Was expecting something more like '%s-step-%05d' % (_cleanse_gcp_job_id(self._job_key), step_num)